Add `MaterialStateColor` support to `TabBar.labelColor`

werainkhatri opened the pull request:
Brings `MaterialStateColor` support to  `TabBar.labelColor` and document why `TabBar.labelStyle.color` and `TabBarTheme.labelStyle.color` cannot influence tab's label color.

### Previous Aim:
Enables `TabBar.labelStyle.color` (and `TabBar.unselectedLabelStyle.color`) to influence the color of the tab's label by adding it (and `TabBarTheme.labelStyle.color`) to the precedence calculation.

Fixes (no longer) #109484 due to https://github.com/flutter/flutter/pull/109541#issuecomment-1294241417

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

werainkhatri committed:
added MaterialStateColor support to TabBarTheme.labelColor

werainkhatri commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -735,19 +739,23 @@ class TabBar extends StatefulWidget implements PreferredSizeWidget {
   ///
-  /// If this parameter is null, then the color of the [ThemeData.primaryTextTheme]'s
-  /// bodyText1 text color is used.
+  /// If null, then [labelStyle.color] will be used. If that is null, then
```
```suggestion
  /// If null, then [labelStyle]'s color is used. If that is null, then
```

werainkhatri commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -735,19 +739,23 @@ class TabBar extends StatefulWidget implements PreferredSizeWidget {
   ///
-  /// If this property is null, unselected tab labels are rendered with the
-  /// [labelColor] with 70% opacity.
+  /// If null, then [unselectedLabelStyle.color] will be used. If that is null,
```
```suggestion
  /// If null, then [unselectedLabelStyle]'s color is used. If that is null,
```

HansMuller commented:
Is this a good opportunity to add support for MaterialStateColor?

werainkhatri commented:
> Is this a good opportunity to add support for MaterialStateColor?

only `labelColor` and `labelStyle` should support `MaterialStateColor`, right?

HansMuller commented on packages/flutter/lib/src/material/tabs.dart:
```diff
@@ -174,17 +174,50 @@ class _TabStyle extends AnimatedWidget {
+        ?? tabBarTheme.labelStyle?.color
+        ?? themeData.primaryTextTheme.bodyLarge!.color!;
+
+    if (selectedColor is MaterialStateColor) {
```
This is good, in that it supports a MaterialStateColor from the widget or the theme, however it doesn't support lerping between the selected and unselected colors if the labelColor is specified as a MaterialStateColor. This could be corrected here, by resolving the MaterialState selectedColor against `<MaterialState>{MaterialState.selected}` and `<MaterialState>{}` and using those two colors in the resolveWith expression below.

werainkhatri replied:
Oh right, thanks for catching that. will add a test for it too.

HansMuller commented:
This is failing internal tests of apps that specify TabBar.labelStyle but appear to rely on the old behavior which ignored the labelStyle's color when computing the selectedColor:

```dart
    final Color selectedColor = labelColor
       ?? tabBarTheme.labelColor
       ?? themeData.primaryTextTheme.bodyLarge!.color!;
```

Unfortunately, this is a breaking change, and not one where it would be possible to automatically migrate existing apps.

werainkhatri commented:
oh that's unfortunate. i went ahead and removed it. updated title and description to reflect support for `MaterialStateColor` along with documenting why this was done. hope that is sufficient.

just thinking out loud: if this is (or becomes) a high customer request, would it make sense to add a variable to decide if `labelStyle` should affect `labelColor`, defaulting it to false? or would it make sense to introduce this breaking change as one can't think of a usecase where this is behaviour expected, unless done unintentionally. thoughts?

HansMuller commented:
> just thinking out loud: if this is (or becomes) a high customer request, would it make sense to add a variable to decide if labelStyle should affect labelColor, defaulting it to false? or would it make sense to introduce this breaking change as one can't think of a usecase where this is behaviour expected, unless done unintentionally. thoughts?

Something like that would be needed to safely migrate existing applications, yes. As you can imagine, it's quite a production, so - as you pointed out - we'd want to wait until there was strong demand for the feature.

werainkhatri commented:
got it.

at this time, MaterailStateColor changes are suffitient right? could you or @Renzo-Olivares review it?

thanks!

HansMuller commented:
The MaterialStateColor changes are sufficient. Just waiting for "Google Testing" to give its thumbs-up.

werainkhatri commented:
Hey @HansMuller, could you take a look? Google testing is failing again. 😬

HansMuller commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -3950,6 +3950,141 @@ void main() {
+  testWidgets('TabBar colors labels correctly', (WidgetTester tester) async {
+    int color = 0x00000000;
+    MaterialStateColor buildMSC() {
+      final Color selectedColor = Color(color++);
```
This isn't wrong but we typically don't write this kind of thing in the Flutter SDK. Creating a bunch of const color and text styles is only slightly more tedious (and a little more Flutter-conventional). And then buildMSC() can just have a pair of parameters.

werainkhatri replied:
i just didn't want the next guy who may have to add a few more colors to test a few more cases to keep track of the numbers. will fix it anyway.

HansMuller commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -3950,6 +3950,141 @@ void main() {
+      unselectedLabelStyle: TextStyle(color: Color(color++)),
+    );
+
+    Future<void> buildTabBar({
```
This is OK, bug typically the pumpWidget call is inline; `await test.pumpWidget(buildTabBar(...))`

werainkhatri replied:
hmm that makes sense.

HansMuller commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -3950,6 +3950,141 @@ void main() {
+      ));
+    }
+
+    Color? getTab1Color() => IconTheme.of(tester.element(find.text('1'))).color;
```
Comparing colors for == can be tricky because of MaterialColor etc. Could defend against that (future problem) by returning `.color.value` here and below. Then:
```dart
expect(getTab1Color(), getSelectedColor(materialLabelColor))
```

werainkhatri replied:
ohh. i didn't think of that. thanks!

HansMuller commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -3950,6 +3950,141 @@ void main() {
+    Color getUnselectedColor(Color color) => (color as MaterialStateColor)
+        .resolve(<MaterialState>{});
+
+    // highest precedence: labelColor as MaterialStateColor
```
It's great to see that all the override cases are checked here.

HansMuller commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -4318,46 +4454,82 @@ void main() {
-              return const Text('Second');
-            }),
-          ],
+    Future<void> buildTabBar({bool labelColorIsMaterialStateColor = false}) async {
```
Same comment about conventional factoring and pumpWidget() as above.

HansMuller commented on packages/flutter/test/material/tabs_test.dart:
```diff
@@ -4309,7 +4444,8 @@ void main() {
 
-  testWidgets('TabController changes offset attribute', (WidgetTester tester) async {
+  testWidgets('TabController.offset changes reflect labelColor',
+          (WidgetTester tester) async {
```
no need for a newline here

werainkhatri replied:
it was passing the 80 character limit, that's why I did it. will revert it sure.

HansMuller reviewed (commented):
This looks good, just some nitpicks about the new tests.

HansMuller reviewed (approved):
LGTM
